pandas/core/arrays/arrow/array.py

```diff
@@ -120,19 +120,54 @@ def cast_for_truediv(
             # Intentionally not using arrow_array.cast because it could be a scalar
             # value in reflected case, and safe=False only added to
             # scalar cast in pyarrow 13.
-            return pc.cast(arrow_array, pa.float64(), safe=False)
-        return arrow_array
+            # In arrow, common type between integral and float64 is float64,
+            # but integral type is safe casted to float64, to mirror python
+            # and numpy, we want an unsafe cast, so we cast both operands to
+            # to float64 before invoking arrow.
+            return pc.cast(arrow_array, pa.float64(), safe=False), pc.cast(
+                pa_object, pa.float64(), safe=False
+            )
+
+        return arrow_array, pa_object
 
     def floordiv_compat(
         left: pa.ChunkedArray | pa.Array | pa.Scalar,
         right: pa.ChunkedArray | pa.Array | pa.Scalar,
     ) -> pa.ChunkedArray:
-        # Ensure int // int -> int mirroring Python/Numpy behavior
-        # as pc.floor(pc.divide_checked(int, int)) -> float
-        converted_left = cast_for_truediv(left, right)
-        result = pc.floor(pc.divide(converted_left, right))
         if pa.types.is_integer(left.type) and pa.types.is_integer(right.type):
+            # Use divide_checked to ensure cases like -9223372036854775808 // -1
+            # don't silently overflow.
+            divided = pc.divide_checked(left, right)
+            # GH 56676: avoid storing intermediate calculating in floating point type.
+            has_remainder = pc.not_equal(pc.multiply(divided, right), left)
+            result = pc.if_else(
+                # Pass a typed arrow scalar rather than stdlib int
+                # which always inferred as int64, to prevent overflow
+                # in case of large uint64 values.
+                pc.and_(
+                    pc.less(
+                        pc.bit_wise_xor(left, right), pa.scalar(0, type=divided.type)
```

Comment:
Does this condition not break down when dealing with unsigned operands?

Reply:
It should evaluate to false, test_arrow_floordiv_no_overflow covers unsigned operands.

Reply:
If both operants are unsigned, then this operation could be optimized to just return pc.divide_checked since integer division for positive values should be same as floor division. Will make this change.

Reply:
Ah OK. I guess for unsigned arguments the else branch would always occur here

Reply:
Yes, I've improved the code to avoid doing any correction for unsigned types, for only positive values, floor division and integer division should be the same.